Drv/LinuxUartDriver/LinuxUartDriver.cpp

```diff
@@ -0,0 +1,404 @@
+    (void)tcflush(fd, TCIFLUSH);
+
+    // Set attributes:
+    stat = tcsetattr(fd, TCSANOW, &newtio);
+    if (-1 == stat) {
+        DEBUG_PRINT("tcsetattr UART fd %d failed\n", fd);
+        close(fd);
+        Fw::LogStringArg _arg = device;
+        Fw::LogStringArg _err = strerror(errno);
+        this->log_WARNING_HI_DR_OpenError(_arg, fd, _err);
+        return false;
+    }
+
+    // All done!
+    Fw::LogStringArg _arg = device;
+    this->log_ACTIVITY_HI_DR_PortOpened(_arg);
+    this->ready_out(0); // Indicate the driver is connected
+    return true;
+}
+
+LinuxUartDriver ::~LinuxUartDriver() {
+    if (this->m_fd != -1) {
+        DEBUG_PRINT("Closing UART device %d\n", this->m_fd);
+
+        (void)close(this->m_fd);
+    }
+}
+
+// ----------------------------------------------------------------------
+// Handler implementations for user-defined typed input ports
+// ----------------------------------------------------------------------
+
+Drv::SendStatus LinuxUartDriver ::send_handler(const NATIVE_INT_TYPE portNum, Fw::Buffer& serBuffer) {
+    if (this->m_fd == -1) {
+        return Drv::SendStatus::SEND_ERROR;
+    }
+
+    unsigned char* data = serBuffer.getData();
+    NATIVE_INT_TYPE xferSize = serBuffer.getSize();
+
+    NATIVE_INT_TYPE stat = ::write(this->m_fd, data, xferSize);
+
+    // TODO no need to delay for writes b/c the write blocks
+    // not sure if it will block until everything is transmitted, but seems to
+
+    if (-1 == stat || stat != xferSize) {
+        Fw::LogStringArg _arg = this->m_device;
+        this->log_WARNING_HI_DR_WriteError(_arg, stat);
+        return Drv::SendStatus::SEND_ERROR;
+    }
+    return Drv::SendStatus::SEND_OK;
+}
+
+void LinuxUartDriver ::serialReadTaskEntry(void* ptr) {
+    Drv::RecvStatus status = RecvStatus::RECV_ERROR;  // added by m.chase 03.06.2017
+
+    LinuxUartDriver* comp = static_cast<LinuxUartDriver*>(ptr);
+
+    while (true) {
+        // wait for data
+        int sizeRead = 0;
+
+        Fw::Buffer buff = comp->allocate_out(0, Drv::UART_READ_ALLOCATION_REQUEST_SIZE);
+
+
+        if (buff.getData() == nullptr) {
+            Fw::LogStringArg _arg = comp->m_device;
+            comp->log_WARNING_HI_DR_NoBuffers(_arg);
+            comp->recv_out(0, buff, status);
+            // to avoid spinning, wait 50 ms
+            Os::Task::delay(50);
+            continue;
+        }
+
+        //          timespec stime;
+        //          (void)clock_gettime(CLOCK_REALTIME,&stime);
+        //          DEBUG_PRINT("<<< Calling dsp_relay_uart_relay_read() at %d %d\n", stime.tv_sec, stime.tv_nsec);
+
+        bool waiting = true;
+        int stat = 0;
+
+        while (waiting) {
+            if (comp->m_quitReadThread) {
+                return;
+            }
+
+            stat = ::read(comp->m_fd, buff.getData(), buff.getSize());
+
+            // Good read:
+            if (stat > 0) {
+                sizeRead = stat;
+                // TODO remove
```

Comment:
Still needed?

Reply:
I removed a lot of code there.  This was don't to prevent returning/breaking when not returning the allocated buffer.